### DISCUSSION TITLE
Shader usage help

### DISCUSSION THREAD

**Discussion by marc2332** [rust-skia/rust-skia#800] [Q&A]
hey Armin!

I am trying to figure out how to load a "fragment shader". I only have an issue with setting `uniform` values, I can't seem to find a way to do it, there is no `set` or anything similar in the `Data` type. [Here](https://fiddle.skia.org/c/16cadeb67ec0854758dc044a2bd6091c) is a fiddle with the cpp skia, rendering a shader

```rust
let shader = RuntimeEffect::make_for_shader(SHADER, None).unwrap();

let uniforms = Data::new_empty();

// uniforms.set("iTime", 0.5);

let shader = shader.make_shader(uniforms, &[], None).unwrap();

...etc
```

Could you point me to the right direction? I am new to shaders

Thank you again :D

**Comment by pragmatrix** [comment]
Hey Marc,

while I have used a few shaders before, I've never used shaders in Skia. 

The bindings are incomplete, because SkSL Shader support was recently introduced and the comments state that some of the new classes are experimental, specifically the new builders you seem to need for providing uniforms.

To work around that, you can get the offsets and types that need to be provided with SkData by using the `RuntimeEffect::find_uniform` function (that's what the builders use).

I'll see if I find some time to add bindings for the builders.

**Reply by marc2332** [reply]
Awesome, I'll see what I can do with that

Sorry for creating an issue instead of a Q&A discussion :pray:

**Reply by marc2332** [reply]
Happy to report they work :)) I just had to pass the uniform values as you told me (still not sure about my implementation tho, but it seems to work!)

https://user-images.githubusercontent.com/38158676/233185958-e8c3f67e-aed4-4e46-afa9-97d3b1eac311.mp4
